### ISSUE TITLE
[BUG] tag_build breaks the use of data files

### ISSUE THREAD

**Issue by eli-schwartz** [pypa/setuptools#3997]
Environment:
- setuptools 68.0.0
- Python 3.10

setup.cfg:
```cfg
[egg_info]
tag_build = +what
tag_date = 0

[metadata]
name = test
version = 1.0

[options]
packages = find:
package_dir =
    =src

[options.data_files]
share/man/man1 = foo.1
```

setup.py:
```python
from setuptools import setup
setup(headers=['fakehdr.h'])
```

This results in the following wheel contents:

broken:
- test-1.0.data/data/share/man/man1/foo.1
- test-1.0.data/headers/fakehdr.h

working:
- test-1.0+what.dist-info/METADATA
- test-1.0+what.dist-info/WHEEL
- test-1.0+what.dist-info/top_level.txt
- test-1.0+what.dist-info/RECORD 

Notice that tag_build is not included in the naming scheme for files in at least the data/headers schemes. The result is that `python -m installer` fails to detect that these are part of the wheel .data directory, and installs the entire test-1.0.data/ directory as regular package data into site-packages.

It is "correct" to do so since the wheel is broken. But pip has a quirk and regards any directory ending in ".data" as matching for this purpose, so pip install handles this anyway.

This is the root cause for https://github.com/gentoo/gentoo/pull/29795#issuecomment-1508492454

/cc @thesamesam

**Comment by abravalheri** [comment]
Thank you very much @eli-schwartz .
I am having a look. Not sure if this has to be changed directly in https://github.com/pypa/wheel/blob/0.41.0/src/wheel/bdist_wheel.py#L360-L363.

**Comment by abravalheri** [comment]
> If I had to guess, I would say that the version string is set to its final value after `egg_info` runs, so using the version string to set `bdist_wheel.data_dir` before `egg_info` runs will result in the wrong directory name.

Probably we have to run `egg_info.finalize_options()` before attempting to use `distribution.get_version()` to set `bdist_wheel.data_dir`. Reference: https://github.com/pypa/setuptools/blob/v68.0.0/setuptools/command/egg_info.py#L242.

**Comment by eli-schwartz** [comment]
From some intriguingly coincidental timing, someone discovered that `bdist_wheel --build-number XXXX` has the opposite effect... the .data directory has the tag, .dist-info does not. :eyes:

See link above.

**Comment by abravalheri** [comment]
I believe that `--build-number` comes directly from `pypa/wheel` and might not be used in other parts of `pypa/setuptools`.

I could be worth bringing the discussion to `pypa/wheel` and maybe opening a PR with a test case?
Who knows, maybe it is already fixed by pypa/wheel#550

**Comment by abravalheri** [comment]
I will close this issue now because it is likely to have been solved in `wheel==0.41.1`.
